File: src/core/widget.zig
const std = @import("std");
const tui = @import("tui");
const Color = tui.Color;
const tmux_renderer = @import("../tmux_renderer.zig");
const themes = @import("../themes/registry.zig");
const Theme = @import("theme.zig").Theme;

/// Shared context for all widgets that handles common operations:
/// - Theme resolution (built-in or custom)
/// - Transparent background handling
/// - Reset string generation
/// - Color threshold helpers
pub const WidgetContext = struct {
    theme: Theme,
    reset: []const u8,
    allocator: std.mem.Allocator,

    pub fn init(
        allocator: std.mem.Allocator,
        io: std.Io,
        environ_map: *std.process.Environ.Map,
        theme_name: []const u8,
        transparent: bool,
    ) !WidgetContext {
        const theme = (themes.byName(allocator, io, environ_map, theme_name) orelse themes.hard).withTransparentBackground(transparent);

        var buf: [256]u8 = undefined;
        var writer: std.Io.Writer = .fixed(&buf);
        try tmux_renderer.writeReset(theme.foreground, theme.background, &writer);
        const reset = try allocator.dupe(u8, std.Io.Writer.buffered(&writer));

        return WidgetContext{
            .theme = theme,
            .reset = reset,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *WidgetContext) void {
        self.allocator.free(self.reset);
    }

    /// Returns danger for >=80%, warning for >=50%, success otherwise
    pub fn thresholdColor(self: WidgetContext, percent: u8) Color {
        if (percent >= 80) return self.theme.danger;
        if (percent >= 50) return self.theme.warning;
        return self.theme.success;
    }

    /// Returns danger for low battery, success for full, warning otherwise
    pub fn batteryColor(self: WidgetContext, percentage: u8, low_threshold: u8) Color {
        if (percentage < low_threshold) return self.theme.danger;
        if (percentage >= 100) return self.theme.success;
        return self.theme.warning;
    }

    /// Format a status segment with the given color and icon
    pub fn formatSegment(
        self: WidgetContext,
        color: Color,
        icon: []const u8,
        value: usize,
    ) ![]const u8 {
        var fg_buf: [32]u8 = undefined;
        var bg_buf: [32]u8 = undefined;
        return std.fmt.allocPrint(self.allocator, " {s}#[fg={s},bg={s},bold]{s} {d}", .{
            self.reset,
            tmux_renderer.colorHexString(color, &fg_buf),
            tmux_renderer.colorHexString(self.theme.background, &bg_buf),
            icon,
            value,
        });
    }
};

test "WidgetContext thresholdColor" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var ctx = try WidgetContext.init(gpa, io, &env_map, "hard", false);
    defer ctx.deinit();

    try std.testing.expect(std.meta.eql(ctx.theme.danger, ctx.thresholdColor(80)));
    try std.testing.expect(std.meta.eql(ctx.theme.danger, ctx.thresholdColor(100)));
    try std.testing.expect(std.meta.eql(ctx.theme.warning, ctx.thresholdColor(50)));
    try std.testing.expect(std.meta.eql(ctx.theme.warning, ctx.thresholdColor(79)));
    try std.testing.expect(std.meta.eql(ctx.theme.success, ctx.thresholdColor(49)));
    try std.testing.expect(std.meta.eql(ctx.theme.success, ctx.thresholdColor(0)));
}

test "WidgetContext batteryColor" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var ctx = try WidgetContext.init(gpa, io, &env_map, "hard", false);
    defer ctx.deinit();

    try std.testing.expect(std.meta.eql(ctx.theme.danger, ctx.batteryColor(10, 20)));
    try std.testing.expect(std.meta.eql(ctx.theme.warning, ctx.batteryColor(50, 20)));
    try std.testing.expect(std.meta.eql(ctx.theme.success, ctx.batteryColor(100, 20)));
}

test "WidgetContext formatSegment" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var ctx = try WidgetContext.init(gpa, io, &env_map, "hard", false);
    defer ctx.deinit();

    const seg = try ctx.formatSegment(ctx.theme.success, "󰍛", 42);
    defer gpa.free(seg);

    try std.testing.expect(std.mem.containsAtLeast(u8, seg, 1, "#[fg="));
    try std.testing.expect(std.mem.containsAtLeast(u8, seg, 1, "󰍛"));
    try std.testing.expect(std.mem.containsAtLeast(u8, seg, 1, "42"));
}

File: src/core/widget.rs
use crate::core::{Color, Theme};
use crate::themes;
use crate::tmux_renderer::ThemeHex;

#[derive(Clone, Debug)]
pub struct WidgetContext {
    pub theme: Theme,
    pub reset: String,
}

impl WidgetContext {
    pub fn new(theme_name: &str, transparent: bool) -> Self {
        let theme = themes::by_name(theme_name).with_transparent_background(transparent);
        let theme_hex = ThemeHex::from_theme(theme);
        Self::from_theme_hex(theme, &theme_hex)
    }

    pub fn from_theme(theme: Theme) -> Self {
        let theme_hex = ThemeHex::from_theme(theme);
        Self::from_theme_hex(theme, &theme_hex)
    }

    pub fn from_theme_hex(theme: Theme, theme_hex: &ThemeHex) -> Self {
        let reset = format!(
            "#[fg={},bg={},nobold,noitalics,nounderscore,nodim]",
            theme_hex.color(theme.foreground),
            theme_hex.color(theme.background),
        );
        Self { theme, reset }
    }

    pub const fn threshold_color(&self, percent: u8) -> Color {
        if percent >= 80 {
            self.theme.danger
        } else if percent >= 50 {
            self.theme.warning
        } else {
            self.theme.success
        }
    }

    pub const fn battery_color(&self, percentage: u8, low_threshold: u8) -> Color {
        if percentage < low_threshold {
            self.theme.danger
        } else if percentage >= 100 {
            self.theme.success
        } else {
            self.theme.warning
        }
    }

    pub fn format_segment(
        &self,
        theme_hex: &ThemeHex,
        color: Color,
        icon: &str,
        value: usize,
    ) -> String {
        format!(
            " {}#[fg={},bg={},bold]{} {}",
            self.reset,
            theme_hex.color(color),
            theme_hex.color(self.theme.background),
            icon,
            value
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn widget_context_threshold_color() {
        let ctx = WidgetContext::new("hard", false);
        assert_eq!(ctx.theme.danger, ctx.threshold_color(80));
        assert_eq!(ctx.theme.danger, ctx.threshold_color(100));
        assert_eq!(ctx.theme.warning, ctx.threshold_color(50));
        assert_eq!(ctx.theme.warning, ctx.threshold_color(79));
        assert_eq!(ctx.theme.success, ctx.threshold_color(49));
        assert_eq!(ctx.theme.success, ctx.threshold_color(0));
    }

    #[test]
    fn widget_context_battery_color() {
        let ctx = WidgetContext::new("hard", false);
        assert_eq!(ctx.theme.danger, ctx.battery_color(10, 20));
        assert_eq!(ctx.theme.warning, ctx.battery_color(50, 20));
        assert_eq!(ctx.theme.success, ctx.battery_color(100, 20));
    }

    #[test]
    fn widget_context_format_segment() {
        let ctx = WidgetContext::new("hard", false);
        let theme_hex = ThemeHex::from_theme(ctx.theme);
        let segment = ctx.format_segment(&theme_hex, ctx.theme.success, "󰍛", 42);
        assert!(segment.contains("#[fg="));
        assert!(segment.contains("󰍛"));
        assert!(segment.contains("42"));
    }
}
